File: Trees/Java/Delete Nodes And Return Forest.java
// Problem link: https://leetcode.com/problems/delete-nodes-and-return-forest
// Time: O(n)
// Space: O(n)
class Solution {
    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        List<TreeNode> result = new ArrayList<>();
        Set<Integer> deleteSet = new HashSet<>();

        for(int node : to_delete) {
            deleteSet.add(node);
        }
        
        if(this.dfs(root, deleteSet, result) != null) {
            result.add(root);
        }

        return result;
    }

    private TreeNode dfs(TreeNode root, Set<Integer> deleteSet, List<TreeNode> result) {
        if(root == null) {
            return root;
        }

        root.left = this.dfs(root.left, deleteSet, result);
        root.right = this.dfs(root.right, deleteSet, result);

        if(deleteSet.contains(root.val)) {
            if(root.left != null) {
                result.add(root.left);
            }

            if(root.right != null) {
                result.add(root.right);
            }

            return null;
        }

        return root;
    }
}

File: Trees/Python/Delete Nodes And Return Forest.py
# Problem link: https://leetcode.com/problems/delete-nodes-and-return-forest
# Time: O(n)
# Space: O(n)
class Solution:
    def delNodes(self, root: Optional[TreeNode], to_delete: List[int]) -> List[TreeNode]:
        to_delete, result = set(to_delete), []

        if self.dfs(root, to_delete, result):
            result.append(root)

        return result
    
    def dfs(self, root, to_delete, result):
        if not root:
            return None
        
        root.left = self.dfs(root.left, to_delete, result)
        root.right = self.dfs(root.right, to_delete, result)

        if root.val in to_delete:
            if root.left:
                result.append(root.left)
            if root.right:
                result.append(root.right)
            return None
            
        return root
